### PR TITLE
Algo_Registry: Use CRITICAL_SECTION on Windows to preven hang

### PR DESCRIPTION
**Pull request by matejk**
#297: Fix for deadlock in Algo_Registry when loading botan.dll: use critical section instead of std::mutex to prevent deadlock in DllMain when initialising global constants.

### CODE REVIEW

**Comment by randombit** [comment]
Comments:

The preprocessor macro version checks should be only once, and then set a descriptive macro (say BOTAN_WORKAROUND_GH_321 or better referencing a MS bug Id if you can find one) which is used anywhere else it is required.

Make the CS wrapper compatible with std::mutex and std::lock_guard (or at least compatible enough that it compiles with MSVC since that's the only place this will be used). Then have Algo_Registry have a private typedef for the mutex type to use which is #if'ed off the WORKAROUND macro.

**Comment by matejk** [comment]
Proposed improvements are available on the same branch.

**Comment by matejk** [comment]
IMHO it would better in general not to use global variables as initialisers.

An alternative is to either have an initialisation function or to have initialisers as static variables in functions that query for algorithms (like get_cipher).

**Comment by matejk** [comment]
Is modified implementation acceptable for merge?

File: src/lib/base/algo_registry.h [hunk 1]
```diff
@@ -16,8 +16,49 @@
...
+       void unlock()
+          {
+          LeaveCriticalSection(&m_cs);
+          }
+
+    private:
+        CRITICAL_SECTION   m_cs;
+   };
+
+} // namespace
```
Comment (randombit) [inline src/lib/base/algo_registry.h]: Why end and then restart the namespace here? It all could be in a single block

**Comment by randombit** [comment]
Much better, thank you. If you could address the one tiny issue I commented on, and add your copyright info to the header and to doc/license.txt this looks good.

File: src/lib/base/algo_registry.h (hunk 1 above)
Comment (matejk) [inline src/lib/base/algo_registry.h]: There is really no need for that.

**Comment by matejk** [comment]
All of the changes available on the branch.

**Comment by randombit** [comment]
@matejk Looks good, thanks! Will be included in 1.11.25 planned for next week.

**Pull request merged**